index.d.ts

```diff
@@ -1,288 +1,593 @@
+  getRowProps: ((row: Row<D>, instance: TableInstance<D>) => object)[]
+  getHeaderGroupProps: ((
+    headerGroup: HeaderGroup<D>,
+    instance: TableInstance<D>
+  ) => object)[]
+  getHeaderProps: ((column: Column<D>, instance: TableInstance<D>) => object)[]
+  getCellProps: ((cell: Cell<D>, instance: TableInstance<D>) => object)[]
+}
 
-  type StringKey<D> = Extract<keyof D, string>
-  type IdType<D> = StringKey<D> | string
+export interface UseTableColumnOptions<D extends object>
+  extends Accessor<D>,
+    Partial<{
+      columns: Column<D>[]
+      show: boolean | ((instance: TableInstance<D>) => boolean)
+      Header: Renderer<HeaderProps<D>>
+      Cell: Renderer<CellProps<D>>
+    }> {}
+
+export interface UseTableInstanceProps<D extends object> {
+  columns: ColumnInstance<D>[]
+  flatColumns: ColumnInstance<D>[]
+  headerGroups: HeaderGroup<D>[]
+  headers: ColumnInstance<D>[]
+  flatHeaders: ColumnInstance<D>[]
+  rows: Row<D>[]
+  getTableProps: (props?: object) => object
+  prepareRow: (row: Row<D>) => void
+  rowPaths: string[]
+  flatRows: Row<D>[]
+}
 
-  type SortingRule<D> = {
-    id: IdType<D>
-    desc: boolean
-  }
+export interface UseTableHeaderGroupProps<D extends object> {
+  headers: ColumnInstance<D>[]
+  getHeaderGroupProps: (props?: object) => object
+  totalHeaderCount: number
+}
 
-  export type SortingRules<D> = SortingRule<D>[]
+export interface UseTableColumnProps<D extends object> {
+  id: IdType<D>
+  isVisible: boolean
+  render: (type: 'Header' | string, props?: object) => ReactNode
+  getHeaderProps: (props?: object) => object
+  parent: ColumnInstance<D>
+  depth: number
+  index: number
+}
 
-  export type SortByFn = (a: any, b: any, desc: boolean) => 0 | 1 | -1
+export interface UseTableRowProps<D extends object> {
+  cells: Cell<D>[]
+  values: Record<IdType<D>, CellValue>
+  getRowProps: (props?: object) => object
+  index: number
+  original: D
+  path: IdType<D>[]
+  subRows: Row<D>[]
+  state: object
+}
 
-  export type Filters<D> = Record<IdType<D>, string>
+export interface UseTableCellProps<D extends object> {
+  column: ColumnInstance<D>
+  row: Row<D>
+  value: CellValue
+  getCellProps: (props?: object) => object
+  render: (type: 'Cell' | string, userProps?: object) => ReactNode
+}
 
-  export interface Cell<D = {}> extends TableInstance<D> {
-    cell: { value: any }
-    column: Column<D>
-    row: Row<D>
-    render: (type: 'Cell' | 'Aggregated', userProps?: any) => any
-    isGrouped?: boolean
-    isAggregated?: boolean
-    isRepeatedValue?: boolean
-    getCellProps: () => any
-  }
+export type HeaderProps<D extends object> = TableInstance<D> & {
+  column: ColumnInstance<D>
+} & Record<string, any>
+export type CellProps<D extends object> = TableInstance<D> & {
+  column: ColumnInstance<D>
+  row: Row<D>
+  cell: Cell<D>
+} & Record<string, any>
+
+// NOTE: At least one of (id | accessor | Header as string) required
+export interface Accessor<D extends object> {
+  accessor?:
```

Comment:
Before it was clear from these new typings I was using accessor to return an object `x`. Then in cell's `sortType` I was sorting using `x.a` but through `Cell` renderer showing `x.a`.

When cell `accessor` returns only string the table rows can be only sorted by that string that implies alphabetically. For example when I'm displaying formatted money I want to sort it by numerical value that said not by formatted value that servers for display purposes. It would be better to be able to format it in Cell.  
The original behaviour IMO should be kept.

My column example:

```ts
{
    Cell: ({ cell: { value } }) => formatMoney(value),
    accessor: row => somehowExtractMoneyObjectFromRow(row),
    sortType: (a, b, columnId) => {
        const a: Money = a.values[columnId];
        const b: Money = b.values[columnId];

        return Math.sign(a.amount - b.amount) as 1 | 0 | -1;
    }
}

Reply:
If you could provide a quick codesandbox, that would be really helpful! 

Or alternatively, providing an example of your data structure and what you would expect to be the final result vs what you're getting?

Reply:
Here https://codesandbox.io/s/tannerlinsleyreact-table-basic-3o9yy

I display composite value of `firstName` and `age` but sort column only by `age`

Reply:
@simPod I was able to achieve what you wanted with the string accessor.


```js
{
  Header: "Data",
  accessor: 'age',
  Cell: ({ cell: { value }, row: { original: { firstName } }}) => {
    return firstName + " " + value;
  },
  sortType: (aRow, bRow, columnId) => {          
    const a = aRow.values[columnId];
    const b = bRow.values[columnId];
  
    return Math.sign(a - b);
  }
}
```

Reply:
When accessor is a function it's defined as returning the value of the cell, in the same way that row[id] returns the value of a cell.  So while I'm not sure that it should really return an object, I do think that it could reasonably return pretty much any other thing that could be a valid cell value, say number or boolean.

Reply:
Thanks, using CellValue now 👍